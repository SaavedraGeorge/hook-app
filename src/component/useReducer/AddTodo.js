import React from 'react';
import { useForm } from '../../hooks/useForm';

export const AddTodo = ({handleAddTodo}) => {
    const [{description}, handleInputChange, reset] = useForm({ 
        description: ''
     });
    
    const handleSubmit = (e) => {
        e.preventDefault();

        if( description.trim().length <= 1){
            return
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false 
        }

        handleAddTodo(newTodo);
        reset();
    }

    return (
        <>
            <h4>Agregar To Do</h4>
            <hr />
            <form onSubmit={ handleSubmit }>
                <input 
                    type="text" 
                    name="description" 
                    placeholder="add To Do" 
                    autoComplete='off'
                    className='form-control'
                    value={ description }
                    onChange={ handleInputChange }
                />
                <button 
                    className='btn btn-outline-primary col-12 mt-1'
                    type='submit'
                >Agregar</button>
            </form>
        </>
    );
};
