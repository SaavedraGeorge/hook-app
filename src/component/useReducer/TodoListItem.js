import React from 'react';

export const TodoListItem = ({todo, i, handleDelete, toggle }) => {
  
  return (
    <li
    className='list-group-item'
    key={todo.id}
    >
        <p 
            onClick={ () => toggle(todo.id)}
            className={`${ todo.done ? 'complete': ''}`}
        >
            {i+1}. { todo.desc }
        </p>
        <button 
            className='btn btn-danger'
            onClick={ () => handleDelete(todo.id) }
        >Eliminar</button>
    </li>
  );
};
