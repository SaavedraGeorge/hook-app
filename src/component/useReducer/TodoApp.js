import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';


import './style.css'
import { TodoList } from './TodoList';
import { AddTodo } from './AddTodo';

const init = () =>{
    
    // llama lo que hay en el loca storrage y si es vacio devuelve []
    return JSON.parse(localStorage.getItem('todos')) || [];
    
} 

export const TodoApp = () => {

    const [todos, dispatch] = useReducer( todoReducer, [], init );

    
    
    
     useEffect(() => {
       localStorage.setItem('todos', JSON.stringify(todos));
     
     }, [todos]);
     

    const handleAddTodo = ( newTodo ) => {
        
        dispatch({
            type: 'add',
            payload: newTodo
        });

    }

    const handleDelete = (idItem) => {
        
        const action ={
            type: 'delete',
            payload: idItem
        }

        dispatch(action);

    }

    const toggle = (idItem) => {
        dispatch( {
            type: 'toggle',
            payload: idItem 
        } )
    }

    return (
        <div>
            <h1>To Do App <small>{todos.length}</small></h1>
            <hr />

  

            <div className='row'>
                <div className='col-7'>
                    <TodoList 
                        todos = {todos}
                        handleDelete = {handleDelete}
                        toggle = {toggle}
                    /> 
                </div>
                <div className='col-5'>
                    <AddTodo handleAddTodo={handleAddTodo} />
                </div>
            </div>
        </div>
    )
}
