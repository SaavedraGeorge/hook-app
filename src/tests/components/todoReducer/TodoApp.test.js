import React from 'react';
//configure se llama para llamar mount en reat 17
import {mount, shallow, configure} from "enzyme";
//adaptaer es para llamar mount en reat 17
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import { TodoApp } from '../../../component/useReducer/TodoApp';
import { demoTodo } from '../../fixtures/demoTodo';
import { act } from '@testing-library/react';

configure({adapter: new Adapter()});

describe('Pruebas en <TodoApp />', () => {

    //montamos TODOAPP con shallow
    const wrapper = shallow( <TodoApp />);

    //ejecutamos el Storage
    Storage.prototype.setItem = jest.fn(()=> {});

    test('debe de mostrarse correctamente', () => {
        
        expect( wrapper ).toMatchSnapshot();

    });

    test('debe de agregar un TODO', () => {

        
        //se monta el wrapper en mount, es como shallow pero te permite mas opciones
        const wrapper = mount( <TodoApp /> );
        
        //se ejecuta la propiedad (funcion en este paso) handADDTodo desde AddTodo
        //se agregan dos elementos desde demoTodo
         act( () => {
             wrapper.find('AddTodo').prop('handleAddTodo')( demoTodo[0] );
             wrapper.find('AddTodo').prop('handleAddTodo')( demoTodo[1] );
         });

         //venos si los cambios se hicieton primero viendo si el contados que esta en el H1 cambio, en este paso dos veces
         expect( wrapper.find('h1').text().trim() ).toBe('To Do App 2');
         //y se cuenta cuantas veces se llamo al localstorage
         expect( localStorage.setItem ).toHaveBeenCalledTimes(2);
        
    });

    test('debe de eliminar un todo', () => {
        /// la consola nos dira si es neecsario agregar un act

        //se ejecuta la accion handleAddTodo para gregar un elemento
        wrapper.find('AddTodo').prop('handleAddTodo')( demoTodo[0] );
        //se ejecupa ha desde handleDelete desde TodoList
        wrapper.find('TodoList').prop('handleDelete')( demoTodo[0].id );

        //comprobamos con el contador del h1 si el registro se elimino
        expect( wrapper.find('h1').text().trim() ).toBe('To Do App 0');

    });
    
    
    
    
})
