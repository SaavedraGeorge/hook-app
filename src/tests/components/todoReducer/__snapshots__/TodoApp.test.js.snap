// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp /> debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    To Do App 
    <small>
      0
    </small>
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        todos={Array []}
        toggle={[Function]}
      />
    </div>
    <div
      className="col-5"
    >
      <AddTodo
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
